File: shared/filters.py
"""Gorev tablosu filtreleri — taban sinif + turevler (spec/60-kaynak-uyarlama.md 2.2).

Yeni bir boyut eklemek = buraya bir Filter ornegi eklemek; rota ve sablon degismez
(sablon filtreleri genel dongueyle cizer). Abartma: ORM/DSL yok, her filtre tek
WHERE parcasi dondurur. Kurallar spec/10-kararlar.md 'Sorgular':
  - suzme/siralama SQL'de, Python'a donen satir ekranda gorunen satirdir
  - siralama sabit sozlukten (ORDERINGS), kullanici girdisiyle birlestirilmez
  - alt agac tin/tout uzerinden bellekteki agactan, recursive CTE yok
  - metin aramasi tsvector/GIN, LIKE '%..%' yok
"""
from __future__ import annotations

from datetime import datetime, timedelta, timezone

from . import db, search, service

# Sorgular items'i "i" takma adiyla kullanir; tum clause'lar buna gore yazilir.

ORDERINGS = {  # anahtar disaridan gelir ama SQL sabit buradan okunur
    "activity": "i.updated_at desc",
    "date":     "i.due_date is null, i.due_date",
    "priority": ("case i.priority when 'critical' then 0 when 'high' then 1"
                 " when 'medium' then 2 else 3 end"),
    "newest":   "i.created_at desc",
}
DEFAULT_ORDER = "activity"

OPEN_ACTION = "select item_id from actions where status in ('open','in_progress')"


class Filter:
    """Taban sinif. options() sablondaki select'i besler, clause() SQL uretir."""

    def __init__(self, param: str, label: str):
        self.param, self.label = param, label

    def options(self) -> list[tuple[str, str, str | None]]:
        """[(deger, etiket, grup)] — grup None ise optgroup acilmaz."""
        return []

    def clause(self, value: str, user) -> tuple[str, list] | None:
        raise NotImplementedError


class SelectFilter(Filter):
    """Sabit sozluklu sutun esitligi: tur, durum, oncelik."""

    def __init__(self, param: str, label: str, column: str, choices: dict[str, str]):
        super().__init__(param, label)
        self.column, self.choices = column, choices

    def options(self):
        return [(k, v, None) for k, v in self.choices.items()]

    def clause(self, value, user):
        if value not in self.choices:
            return None
        return f"i.{self.column} = %s", [value]


class PersonFilter(Filter):
    """Kullanici sutunu; 'ben' ve 'yok' ozel degerleri."""

    def __init__(self, param: str, label: str, column: str):
        super().__init__(param, label)
        self.column = column

    def options(self):
        special = [("me", "Ben", None), ("none", "Atanmamış", None)]
        return special + [(u["id"], u["name"], None) for u in db.q("select id,name from users order by name")]

    def clause(self, value, user):
        if value == "me":
            return f"i.{self.column} = %s", [user["id"]]
        if value == "none":
            return f"i.{self.column} is null", []
        # Sutun uuid: gecersiz metin dogrudan sorguya girerse veritabani hata
        # verir. Once cevir, olmuyorsa filtreyi yok say (bozuk girdi filtreyi
        # dusurur, istegi dusurmez).
        id_ = db.uid(value)
        if id_ is None or db.q1("select 1 from users where id = %s", (id_,)) is None:
            return None
        return f"i.{self.column} = %s", [id_]


class TeamFilter(Filter):
    def options(self):
        return [(t["id"], t["name"], None) for t in db.q("select id,name from teams order by name")]

    def clause(self, value, user):
        id_ = db.uid(value)
        if id_ is None or db.q1("select 1 from teams where id = %s", (id_,)) is None:
            return None
        return "i.team_id = %s", [id_]


class NodeFilter(Filter):
    """Alt agac suzmesi. Secenekler bellekteki agactan, veri yonetiminde tanimlanan
    turlere (node_type) gore gruplanir — sema degisince filtre kendiliginden uyar."""

    def options(self):
        tree = service.TREE
        ordered = sorted(tree.nodes, key=lambda n: tree.tin[n])
        return [(nid, "· " * tree.depth[nid] + tree.name(nid), tree.nodes[nid].node_type)
                for nid in ordered]

    def clause(self, value, user):
        # subtree() bellekteki agactan; anahtar uuid, gelen deger metin.
        ids = service.TREE.subtree(db.uid(value)) if db.uid(value) else []
        if not ids:
            return None
        return "i.node_id = any(%s)", [list(ids)]


class SearchFilter(Filter):
    """tsvector/GIN — sorgu ifadesi kullanici metniyle birlestirilmez (shared/search.py)."""

    def options(self):
        return []          # select degil metin girisi; sablon bunu options() bos diye anlar

    def clause(self, value, user):
        match = search.fts_query(value)
        if match is None:
            return None
        return "i.search_vector @@ to_tsquery('tr', %s)", [match]


def active_filters() -> list[Filter]:
    """Her istekte kurulur: dugumler agactan gelir."""
    return [
        SelectFilter("kind", "Tür", "kind", {"issue": "Hata", "task": "Görev"}),
        SelectFilter("status", "Durum", "status", dict(service.STATUSES)),
        SelectFilter("priority", "Öncelik", "priority", dict(service.PRIORITIES)),
        TeamFilter("team", "Takım"),
        PersonFilter("person", "Sorumlu", "assignee_id"),
        NodeFilter("node", "Düğüm"),
        SearchFilter("search", "Ara"),
    ]


# --- hizli filtreler: kadans hafta (spec/10-kararlar.md 'Kadans hafta') ------

def _week_range():
    """Tarihler ARTIK METIN DEGIL: sutunlar date/timestamptz, karsilastirma
    icin gercek nesne gonderilir (isoformat() SQLite doneminden kalmaydi)."""
    today = datetime.now(timezone.utc).date()
    return today - timedelta(days=7), today + timedelta(days=7)


def quick_clause(key: str, user) -> tuple[str, list] | None:
    before, after = _week_range()
    today = datetime.now(timezone.utc).date()
    Q = {
        # bu haftanin gundemi: son 7 gunde hareket VEYA son tarihi 7 gun icinde
        "week": ("(i.updated_at >= %s::date or (i.due_date is not null and i.due_date <= %s"
                 " and i.status <> 'closed'))", [before, after]),
        # acik eylemim: actions tablosundan (spec/20-sema.md §3a)
        "my_actions": (f"i.id in (select item_id from actions where assignee_id = %s"
                       " and status in ('open','in_progress'))", [user["id"]]),
        # geciken: kaydin ya da acik bir eyleminin son tarihi gecmis
        "overdue": ("((i.due_date < %s and i.status <> 'closed') or i.id in"
                    " (select item_id from actions where due_date < %s"
                    "  and status in ('open','in_progress')))", [today, today]),
        "unassigned": ("i.assignee_id is null and i.status <> 'closed'", []),
    }
    return Q.get(key)


QUICK_FILTERS = [("all", "Hepsi"), ("week", "Bu hafta"), ("my_actions", "Açık eylemim"),
                 ("overdue", "Geciken"), ("unassigned", "Atanmamış")]


def build_query(params, user) -> tuple[str, list, str, dict]:
    """Istek parametrelerinden (where, args, order by, secili) uretir.

    selected: sablonun select'leri isaretlemesi icin {param: deger} — yalnizca
    gecerli clause ureten degerler girer, yansitilan ham girdi degil.
    """
    where, args, selected = ["1=1"], [], {}
    for f in active_filters():
        v = (params.get(f.param) or "").strip()
        if not v:
            continue
        c = f.clause(v, user)
        if c is None:
            continue
        where.append(c[0])
        args.extend(c[1])
        selected[f.param] = v

    quick = params.get("quick") or "all"
    c = quick_clause(quick, user)
    if c is not None:
        where.append(c[0])
        args.extend(c[1])
    else:
        quick = "all"
    selected["quick"] = quick

    sort = params.get("sort") if params.get("sort") in ORDERINGS else DEFAULT_ORDER
    selected["sort"] = sort
    # deterministik kuyruk: (secilen sutun, id) — spec/10-kararlar.md
    return " and ".join(where), args, f"{ORDERINGS[sort]}, i.id", selected

File: backend/src/db/filters.rs
//! Gorev tablosu filtreleri.
//!
//! Kurallar (spec/10-kararlar.md "Sorgular"):
//!   - suzme/siralama SQL'de; Rust'a donen satir EKRANDA GORUNEN satirdir
//!   - siralama SABIT sozlukten okunur, kullanici girdisiyle birlestirilmez
//!   - alt agac tin/tout ile bellekteki agactan — recursive CTE yok
//!   - metin aramasi tsvector/GIN, LIKE '%..%' yok
//!
//! GECERSIZ DEGER FILTREYI DUSURUR, ISTEGI DEGIL. `?team=xx` ya da
//! `?sort='; drop table records;--` sessizce yok sayilir ve tablo suzmesiz
//! doner — 400 vermek kullaniciya bir sey kazandirmiyor, sorguya sizmak ise
//! kaybettiriyor.

use std::collections::HashMap;

use sqlx::{Postgres, QueryBuilder};
use uuid::Uuid;

use crate::{db::tree::TreeIndex, models::user::User};

/// Anahtar disaridan gelir ama SQL SABIT buradan okunur — birlestirme yok.
const ORDERINGS: &[(&str, &str)] = &[
    ("activity", "r.updated_at desc"),
    ("date", "r.due_date is null, r.due_date"),
    ("priority", "case r.priority when 'critical' then 0 when 'high' then 1 \
                  when 'medium' then 2 else 3 end"),
    ("newest", "r.created_at desc"),
];
const DEFAULT_ORDER: &str = "activity";

pub const QUICK_FILTERS: &[(&str, &str)] = &[
    ("all", "Hepsi"), ("week", "Bu hafta"), ("my_actions", "Açık eylemim"),
    ("overdue", "Geciken"), ("unassigned", "Atanmamış"),
];

const KINDS: &[&str] = &["issue", "task"];
const STATUSES: &[&str] = &["open", "in_progress", "pending", "closed"];
const PRIORITIES: &[&str] = &["critical", "high", "medium", "low"];

#[derive(Debug, Clone, PartialEq)]
pub enum Person { Me, None_, Id(Uuid) }

#[derive(Debug, Clone, PartialEq)]
pub enum Pillar { None_, Id(Uuid) }

#[derive(Debug, Default)]
pub struct Filters {
    pub kind: Option<String>,
    pub status: Option<String>,
    pub priority: Option<String>,
    pub team: Option<Uuid>,
    pub person: Option<Person>,
    pub node: Option<Uuid>,
    pub pillar: Option<Pillar>,
    pub search: Option<String>,
    pub quick: String,
    pub sort: String,
}

fn pick(p: &HashMap<String, String>, key: &str, allowed: &[&str]) -> Option<String> {
    p.get(key).map(|s| s.trim()).filter(|s| allowed.contains(s)).map(String::from)
}

fn uuid_of(p: &HashMap<String, String>, key: &str) -> Option<Uuid> {
    p.get(key).and_then(|s| s.trim().parse().ok())
}

impl Filters {
    /// Ham sorgu parametrelerinden. Her alan KENDI BASINA dogrulanir; biri
    /// gecersizse yalniz o duser.
    pub fn parse(p: &HashMap<String, String>) -> Self {
        Filters {
            kind: pick(p, "kind", KINDS),
            status: pick(p, "status", STATUSES),
            priority: pick(p, "priority", PRIORITIES),
            team: uuid_of(p, "team"),
            person: p.get("person").map(|s| s.trim()).and_then(|v| match v {
                "" => None,
                "me" => Some(Person::Me),
                "none" => Some(Person::None_),
                other => other.parse().ok().map(Person::Id),
            }),
            node: uuid_of(p, "node"),
            pillar: p.get("pillar").map(|s| s.trim()).and_then(|v| match v {
                "" => None,
                "none" => Some(Pillar::None_),
                other => other.parse().ok().map(Pillar::Id),
            }),
            search: p.get("search").map(|s| s.trim().to_string()).filter(|s| !s.is_empty()),
            quick: p.get("quick").map(|s| s.trim()).filter(|q| {
                QUICK_FILTERS.iter().any(|(k, _)| k == q)
            }).unwrap_or("all").to_string(),
            sort: p.get("sort").map(|s| s.trim()).filter(|s| {
                ORDERINGS.iter().any(|(k, _)| k == s)
            }).unwrap_or(DEFAULT_ORDER).to_string(),
        }
    }

    /// `where` parcalarini ekler. Sorgu `records r` takma adini kullanir.
    pub fn push_where(&self, qb: &mut QueryBuilder<Postgres>, user: &User, tree: &TreeIndex) {
        qb.push(" where 1=1");

        if let Some(v) = &self.kind { qb.push(" and r.kind = ").push_bind(v.clone()); }
        if let Some(v) = &self.status { qb.push(" and r.status = ").push_bind(v.clone()); }
        if let Some(v) = &self.priority { qb.push(" and r.priority = ").push_bind(v.clone()); }
        if let Some(v) = self.team { qb.push(" and r.team_id = ").push_bind(v); }

        match &self.person {
            Some(Person::Me) => { qb.push(" and r.owner_id = ").push_bind(user.id); }
            Some(Person::None_) => { qb.push(" and r.owner_id is null"); }
            Some(Person::Id(id)) => { qb.push(" and r.owner_id = ").push_bind(*id); }
            None => {}
        }

        // ALT AGAC bellekteki agactan — Euler araligi, recursive CTE yok.
        // Bilinmeyen dugum bos kume verir ve filtre DUSER (istegi dusurmez).
        if let Some(id) = self.node {
            let ids = tree.subtree(id);
            if !ids.is_empty() {
                qb.push(" and r.unit_id = any(").push_bind(ids).push(")");
            }
        }

        match &self.pillar {
            Some(Pillar::None_) => { qb.push(" and r.pillar_id is null"); }
            Some(Pillar::Id(id)) if tree.get(*id).is_some() => {
                qb.push(" and r.pillar_id = ").push_bind(*id);
            }
            _ => {}
        }

        if let Some(q) = &self.search {
            // tsvector/GIN. Kullanici metni HAM sorguya degil, kelime basi
            // prefix'ine cevrilir (KNOW-145'in Postgres karsiligi).
            let terms: Vec<String> = q.split_whitespace()
                .filter(|w| !w.is_empty())
                .map(|w| format!("{}:*", w.replace(['&', '|', '!', '(', ')', ':', '\''], " ").trim()))
                .filter(|w| w.len() > 2)
                .collect();
            if !terms.is_empty() {
                qb.push(" and r.search_vector @@ to_tsquery('tr', ")
                  .push_bind(terms.join(" & ")).push(")");
            }
        }

        self.push_quick(qb, user);
    }

    /// Hizli filtreler — kadans HAFTA (spec/10-kararlar.md).
    fn push_quick(&self, qb: &mut QueryBuilder<Postgres>, user: &User) {
        let today = chrono::Utc::now().date_naive();
        match self.quick.as_str() {
            // Bu haftanin gundemi: son 7 gunde hareket VEYA son tarihi 7 gun icinde.
            "week" => {
                qb.push(" and (r.updated_at >= ")
                  .push_bind(today - chrono::Duration::days(7))
                  .push(" or (r.due_date is not null and r.due_date <= ")
                  .push_bind(today + chrono::Duration::days(7))
                  .push(" and r.status <> 'closed'))");
            }
            "my_actions" => {
                qb.push(" and r.id in (select record_id from actions where owner_id = ")
                  .push_bind(user.id)
                  .push(" and status in ('open','in_progress'))");
            }
            // GECIKEN: kaydin KENDISI degil, acik bir EYLEMININ son tarihi
            // gecmis olsa da dusmeli.
            "overdue" => {
                qb.push(" and ((r.due_date < ").push_bind(today)
                  .push(" and r.status <> 'closed') or r.id in \
                         (select record_id from actions where due_date < ")
                  .push_bind(today)
                  .push(" and status in ('open','in_progress')))");
            }
            "unassigned" => { qb.push(" and r.owner_id is null and r.status <> 'closed'"); }
            _ => {}
        }
    }

    /// Deterministik kuyruk: (secilen sutun, id). Ayni veride ayni sira.
    pub fn order_by(&self) -> String {
        let col = ORDERINGS.iter().find(|(k, _)| *k == self.sort)
            .map(|(_, v)| *v)
            .unwrap_or_else(|| ORDERINGS[0].1);
        format!(" order by {col}, r.id")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn params(pairs: &[(&str, &str)]) -> HashMap<String, String> {
        pairs.iter().map(|(k, v)| (k.to_string(), v.to_string())).collect()
    }

    #[test]
    fn gecersiz_deger_filtreyi_dusurur_istegi_degil() {
        let f = Filters::parse(&params(&[
            ("team", "xx"), ("sort", "'; drop table records;--"), ("quick", "bilinmez"),
            ("status", "uydurma"), ("kind", "hicbiri"),
        ]));
        assert!(f.team.is_none(), "gecersiz uuid duser");
        assert!(f.status.is_none() && f.kind.is_none(), "sozlukte olmayan deger duser");
        assert_eq!(f.sort, "activity", "gecersiz sort VARSAYILANA duser");
        assert_eq!(f.quick, "all", "gecersiz quick VARSAYILANA duser");
    }

    #[test]
    fn siralama_sabit_sozlukten() {
        // Kullanici girdisi order by'a HIC girmez: bilinmeyen anahtar zaten
        // parse'ta duser, order_by sabit metinden okur.
        let f = Filters::parse(&params(&[("sort", "'; drop table records;--")]));
        let o = f.order_by();
        assert!(!o.contains("drop"), "kullanici girdisi order by'a sizmamali");
        assert!(o.ends_with(", r.id"), "deterministik kuyruk");
    }

    #[test]
    fn gecerli_degerler_gecer() {
        let f = Filters::parse(&params(&[
            ("kind", "issue"), ("status", "open"), ("priority", "critical"),
            ("person", "me"), ("pillar", "none"), ("quick", "overdue"), ("sort", "newest"),
        ]));
        assert_eq!(f.kind.as_deref(), Some("issue"));
        assert_eq!(f.person, Some(Person::Me));
        assert_eq!(f.pillar, Some(Pillar::None_));
        assert_eq!(f.quick, "overdue");
        assert!(f.order_by().contains("created_at desc"));
    }
}
